inject/src/main/java/io/micronaut/context/BeanProvider.java

```diff
@@ -16,19 +16,101 @@
 package io.micronaut.context;
 
 import io.micronaut.core.annotation.NonNull;
+import io.micronaut.core.type.Argument;
+
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.Objects;
+import java.util.function.Consumer;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
 
 /**
  * A provider contract.
  *
  * @param <T> The type
  * @author James Kleeh
+ * @author graemerocher
  * @since 2.4.0
  */
-public interface BeanProvider<T> {
+public interface BeanProvider<T> extends Iterable<T> {
 
     /**
      * @return A fully-constructed and injected instance of T.
      */
     @NonNull
     T get();
+
+    @NonNull
+    @Override
+    default Iterator<T> iterator() {
+        return Collections.singletonList(get()).iterator();
+    }
+
+    /**
+     * <p>
+     * When called, provides back a Stream of the beans available in this provider. If no beans are found, it returns an empty
+     * stream.
+     * </p>
+     *
+     * @since 3.0.0
+     * @return a <code>Stream</code> representing the beans associated with this {@link BeanProvider} object
+     */
+    default Stream<T> stream() {
+        return StreamSupport.stream(this.spliterator(), false);
+    }
+
+    /**
+     * <p>Determines if more than one bean matches the specified type and qualifiers.</p>>
+     *
+     * @since 3.0.0
+     * @return <code>true</code> if only one bean matches.
+     */
+    default boolean isUnique() {
+        return true;
+    }
+
+    /**
+     * <p>Determines if there is no bean that matches the required type and qualifiers.</p>
+     *
+     * @since 3.0.0
+     * @return true if no bean matches.
+     */
+    default boolean isPresent() {
+        return true;
+    }
+
+    /**
+     * Is the bean resolvable.
+     * @return True if it is
+     * @since 3.0.0
+     */
+    default boolean isResolvable() {
+        return isUnique() && isPresent();
+    }
+
+    /**
+     * Executes the given logic if the bean is present.
```

Comment:
This should probably explain that in the case of multiple beans only the first one is passed